### ISSUE TITLE
Unresolved directive error message in ManageIQ doc website

### ISSUE THREAD

**Issue by ailisp** [ManageIQ/manageiq_docs#263]
In the installation section, all index.adoc pages (doc-Installing_on_Google_Compute_Engine, doc-Installing_on_Red_Hat_Enterprise_Linux_OpenStack_Platform, etc will cause an error message rendered in the output html: (may be one or more lines similar to these)
```
Unresolved directive in <stdin> - include::topics/installation.adoc[]
Unresolved directive in <stdin> - include::topics/configuration.adoc[]
```
For example, [http://manageiq.org/docs/reference/latest/doc-Installing_on_Google_Compute_Engine/miq/](http://manageiq.org/docs/reference/latest/doc-Installing_on_Google_Compute_Engine/miq/)
I found this may be a bug of asciidoctor and it can't deal with include file in a symbolic links directory with one level upward. For example:
```shell
myfyb@myfyb-ThinkPad-X230:~/redhat/manageiq_docs/doc-Installing_on_Google_Compute_Engine/miq$ ls -l
total 12
lrwxrwxrwx 1 myfyb myfyb  13 Mar  9 10:51 common -> ../../common/
-rw-rw-r-- 1 myfyb myfyb 619 Mar  9 10:51 docinfo.xml
lrwxrwxrwx 1 myfyb myfyb  10 Mar  9 10:51 images -> ../images/
-rw-rw-r-- 1 myfyb myfyb 702 Mar  9 12:17 index.adoc
lrwxrwxrwx 1 myfyb myfyb  10 Mar  9 10:51 topics -> ../topics/
```
In this case, all adoc files in "common" can be correctly included, but all in "topics" can't. I have tried make other symbolic links, such as:
```shell
myfyb@myfyb-ThinkPad-X230:~/redhat/manageiq_docs/doc-Installing_on_Google_Compute_Engine/miq$ mkdir ../../duang
myfyb@myfyb-ThinkPad-X230:~/redhat/manageiq_docs/doc-Installing_on_Google_Compute_Engine/miq$ cp topics/configuration.adoc ../../duang/configuration.adoc
myfyb@myfyb-ThinkPad-X230:~/redhat/manageiq_docs/doc-Installing_on_Google_Compute_Engine/miq$ ln -s ../../duang/ duang
myfyb@myfyb-ThinkPad-X230:~/redhat/manageiq_docs/doc-Installing_on_Google_Compute_Engine/miq$ ls -l
total 12
lrwxrwxrwx 1 myfyb myfyb  13 Mar  9 10:51 common -> ../../common/
-rw-rw-r-- 1 myfyb myfyb 619 Mar  9 10:51 docinfo.xml
lrwxrwxrwx 1 myfyb myfyb  12 Mar  9 12:27 duang -> ../../duang/
lrwxrwxrwx 1 myfyb myfyb  10 Mar  9 10:51 images -> ../images/
-rw-rw-r-- 1 myfyb myfyb 798 Mar  9 12:18 #index.adoc#
-rw-rw-r-- 1 myfyb myfyb 702 Mar  9 12:17 index.adoc
lrwxrwxrwx 1 myfyb myfyb  10 Mar  9 10:51 topics -> ../topics/
```
And in this case, the include line in index.adoc:
```
[[Configuration]]
include::duang/configuration.adoc[]
```
will include as expected. Since "duang" is a meaningless name, and doesn't appear in any config files, **so we know that the asciidoctor can't deal with files in a directory that symbolic linked to one level upward.** And after examine its source, this def is the one asciidoctor to deal with parse include file path, which is buggy:
[https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/abstract_node.rb#L541](https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/abstract_node.rb#L541)
That's also why every .adoc file in ../../common is correctly included but not files in ../topics. I think there are two ways to solve, one is to **fork the asciidoctor and solve it there** (However there are 500+ unsolved issues and many not merged pull requests for asciidoctor so we may have to keep our own fork until it's fixed by official repo). **Another is a dirty fix to change the directory organization of our doc, which will make the doc website include these missing sections correctly but the doc is less organized.** Actually I can do this, as I have fixed part of the doc in this way, but I don't know whether  we should do this way. Which approach should we have? Or a third better solution?

**Comment by ailisp** [comment]
@gtanzillo @jrafanie
I had an interview for a summer intership this Monday. I can't find your public email addresses so I sent a message here, if this annoys you please just ignore it. Thanks for your time to interview with me. Today I have a try on the ManageIQ follows the document. However, it's quite unlucky that most parts on installation and configuration included in the doc just show an error message in the ManageIQ doc website. After clone it source I can actually see the documentation, so they are here but not included correctly. Do you have better ideas on solving this issue? Since missing information for installation and configuration may disappoint potential users. Thank you!

**Comment by jrafanie** [comment]
Thanks @ailisp for the bug report.

@Fryguy What do you think?  Do you have any contacts on the asciidoctor repo?

**Comment by ailisp** [comment]
Sorry, by debugging asciidoctor I just found it's not because asciidoctor
can't deal with symbolic links, but because all the path is based on the
root directory of manageiq's doc. Include files in the "common" directory
happens to included correctly it's because the ../../common relative to
doc-* happens to be manageiq's-doc-root/common. So we need either rewrite
all include lines in index.adoc using path relative to doc root or change
how ascii_binder deal with relative paths.


2017年3月9日 下午1:01，"Joe Rafaniello" <notifications@github.com>写道：

**Comment by ailisp** [comment]
The ascii_binder, unfortunately, unable to deal with the directory included this way, as in ascii_binder's doc it only support 2 level of directories:
[http://www.asciibinder.org/latest/reference/topic_map.html#opinionated-organization](http://www.asciibinder.org/latest/reference/topic_map.html#opinionated-organization)
And it's example in openshift example, the doc is organized like this:
```
- A title (dir)
  - index.adoc
  - subsection1.adoc
  - subsection2.adoc
```
And no include lines in index.adoc, but the ascii_binder includes these files via _topic_map.yml. In manageiq doc, the doc is organized this way:
```
- A title (dir)
  - index.adoc
  - topics (a sym link to related topics dir)
    - subsection1.adoc
    - subsection2.adoc
```
And subsections are included in index.adoc. not under control in ascii_binder but is processed directly by asciidoctor. So here is the problem, we run ascii_binder in the root directory of manageiq_docs, it call asciidoctor, which makes manageiq_docs as the current working directory, with all include files based on this working directory. Without a improve on the ascii_binder project, we have to rearrange manageiq_docs similar way as openshift, or we need to change all include lines in all index.adoc this way:
For example this (manageiq_docs/doc-Installing_on_Google_Compute_Engine/miq/index.adoc):
```
[[Installation]]
include::topics/installation.adoc[]
```
Should be changed to:
```
[[Installation]]
include::doc-Installing_on_Google_Compute_Engine/topics/installation.adoc[]
```
**This is ugly, though works. Better solution should be improve ascii_binder. If you think this is ok, I can send a pull request that modify to make all include works.**

**Comment by adahms** [comment]
@ailisp - Andrew from documentation here, and I have a proposal that may reduce the work required here a little.

At current, the table of contents are split into the master.adoc and index.adoc files, which means we need to update both copies each time the structure of the title changes.

Unless there is a specific need to manage the table of contents separately, why don't we create a 'contents.adoc' file in the 'topics' directory that contains the main includes. This would mean you would only need the full path to the contents.adoc file, and any changes in structure would immediately be reflected in both.

What do you think?

**Comment by ailisp** [comment]
@adahms Thanks, I agree, the table of contents are same in both files, should be make a common one. I also made a local fork of ascii_binder and tried to modify `def configure_and_generate_page` in `ascii_binder/lib/ascii_binder/helpers.rb` to switch to sub directory (e.g. doc-Introduction_to_Self_Service_Portal/ and others) each time process an adoc in a new directory, and switch back to manageiq_docs when done. But I found several files already changed to full path and most are not changed, so when used this version the already modified ones don't work. Shall we use this modified ascii_binder, modify these full paths back to names like "foo.adoc" or  just change all "foo,doc" to full paths?
